velox/functions/prestosql/Sequence.cpp

```diff
@@ -0,0 +1,173 @@
+namespace {
+
+// See documentation at https://prestodb.io/docs/current/functions/array.html
+class SequenceFunction : public exec::VectorFunction {
+ public:
+  static constexpr int32_t kMaxResultEntries = 10'000;
+
+ private:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    VectorPtr localResult;
+    if (args[0]->type() == BIGINT()) {
+      localResult = applyNumber(rows, args, outputType, context);
+    }
+    context.moveOrCopyResult(localResult, rows, result);
+  }
+
+  VectorPtr applyNumber(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& outputType,
+      exec::EvalCtx& context) const {
+    const auto numRows = rows.size();
+    auto pool = context.pool();
+    int totalCount = 0;
+
+    auto startVector = args[0];
+    auto stopVector = args[1];
+    VectorPtr stepVector;
+    if (args.size() == 3) {
+      stepVector = args[2];
+    }
+
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      int count = sequenceCount(startVector, stopVector, stepVector, row);
+      totalCount += count;
+    });
+
+    VectorPtr elements = BaseVector::create(BIGINT(), totalCount, pool);
+    BufferPtr sizes = allocateSizes(numRows, pool);
+    BufferPtr offsets = allocateOffsets(numRows, pool);
+    auto rawElements = elements->asFlatVector<int64_t>()->mutableRawValues();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+
+    int elementsOffset = 0;
+    context.applyToSelectedNoThrow(rows, [&](auto row) {
+      rawOffsets[row] = elementsOffset;
+      rawSizes[row] = sequenceCount(startVector, stopVector, stepVector, row);
+      writeToElements(
+          rawElements,
+          elementsOffset,
+          startVector,
+          stopVector,
+          stepVector,
+          row);
+      elementsOffset += rawSizes[row];
+    });
+
+    return std::make_shared<ArrayVector>(
+        pool, outputType, nullptr, numRows, offsets, sizes, elements);
+  }
+
+  int sequenceCount(
+      VectorPtr startVector,
+      VectorPtr stopVector,
+      VectorPtr stepVector,
+      vector_size_t row) const {
+    long start = valueInVector(startVector, row);
+    long stop = valueInVector(stopVector, row);
+    long step;
+    if (stepVector == nullptr) {
+      step = stop >= start ? 1 : -1;
+    } else {
+      step = valueInVector(stepVector, row);
+    }
+    checkArguments(start, stop, step);
+    return (int)((stop - start) / step + 1);
+  }
+
+  long valueInVector(VectorPtr baseVector, vector_size_t row) const {
+    if (baseVector->isConstantEncoding()) {
+      return baseVector->as<ConstantVector<long long>>()->valueAt(0);
+    } else {
+      return baseVector->asFlatVector<long long>()->valueAt(row);
+    }
+  }
+
+  void checkArguments(int start, int stop, int step) const {
```

Comment:
When a class method does not use any class instance's members it is better to mark it with "static", "const" won't be needed in this case:
```
static void checkArguments(int start, int stop, int step) {
```